mllib/src/main/scala/org/apache/spark/ml/feature/LabeledPoint.scala

```diff
@@ -37,4 +37,9 @@ case class LabeledPoint(@Since("2.0.0") label: Double, @Since("2.0.0") features:
   override def toString: String = {
     s"($label,$features)"
   }
+
+  private[spark] def toInstance(weight: Double): Instance = {
+    Instance(label, weight, features)
```

Comment:
I think this belongs in the Instance class, but, don't feel strongly about it.
It can also specify a default value of 1.0 to avoid writing that in most calls.

Reply:
hmm, interesting, I'm really not sure why we have both LabeledPoint and Instance classes, to me they seem very similar.  In addition, we also have OffsetInstance which has a toInstance method, but no method to convert an Instance to an OffsetInstance or to a LabeledPoint.  It seems it might make more sense to have an Instance(label, features), WeightedInstance(label, features, weight) and OffsetInstance(label, features, weight, offset), at least that would be more consistent in the system and with the terminology.  Or, we could have just a single instance with some optional parameters, but I can see how it would not make sense to waste memory with two optional params.
In terms of where the logic to convert LabeledPoint/Instance should live, I really don't have an opinion.  I think the most important thing is to be consistent, but with these 3 instance/labeledpoint classes I don't see any consistency currently.
However, for this specific PR I think it's more convenient to go from LabeledPoints to Instance objects than to create an Instance from a LabeledPoint.  Isn't calling a method on an object cleaner than having to call a static method?  I agree about the default value though, will make that change.

Reply:
maybe I misunderstood, could you explain a bit more on what the method would look like on the Instance class and how it would be called (and what would need to be imported in each file which uses it)?

Reply:
LabeledPoint is an old class and used everywhere weights don't matter. Terminology inconsistency aside, it makes sense to have a separate class with the weight (rather than carry it around where not needed) and that arose with the newer .ml package. I'd generally prefer to "point" dependencies from .ml to .mllib and not the other way.

I suppose Instance could or should have subclassed LabeledPoint to add weights, and likewise OffsetInstance, so that the conversion is optional one way, but I wouldn't touch that here.

I'd also see if you can minimize the amount of conversion between LabeledPoint and Instance in the test code. It's necessary in most places but we can probably factor out some of it so every caller doesn't have to convert.

Reply:
"I'd generally prefer to "point" dependencies from .ml to .mllib and not the other way."
it seems there are actually two LabeledPoint classes:
/mllib/src/main/scala/org/apache/spark/ml/feature/LabeledPoint.scala
/mllib/src/main/scala/org/apache/spark/mllib/regression/LabeledPoint.scala
I'm not sure why the mllib one is under regression but that is the old one that is used everywhere.
In this PR the toInstance methods are on the new version in ml, not mllib, so we aren't depending from mllib to ml.
However, I could move the method to Instance if you prefer.  I tried locally to add the code below but the changes ended up looking worse than before - I had to use Instance.convert(...) instead of just .toInstance which actually seemed cleaner.  Would it be better for me to add implicit methods then?  But then the code would probably look the same, with a .toInstance, except there would be an extra import probably.  Not sure what would be the better way to handle this.
Here's the conversion method I defined:

private[ml] object Instance {

  /**
   * Convert a LabeledPoint into an Instance.
   * @param labeledPoint LabeledPoint to convert.
   * @param weight Optional weight for the instance.
   * @return Instance representation for this LabeledPoint.
    */
  def convert(labeledPoint: LabeledPoint, weight: Double = 1.0): Instance =
    Instance(labeledPoint.label, weight, labeledPoint.features)
}

Reply:
OK, I'm persuaded. Maybe One Day we can clean up the relationship between these classes a bit more but this isn't particularly worse.